### PR TITLE
add pcre2 support

### LINKED ISSUE #4711: 3.4.6: can't find pcre in /usr/include/pcre
I attempt to build cyrus-imap 3.4.6 with pcre but configure script can;t find it.

$ pkg-config --cflags libpcre
-I/usr/include/pcre

OS: Linux

### PR DESCRIPTION
**Pull request by elliefm**
* Adds pcre2 support based on #4545, but without removing legacy pcre support
* Updates pcre/pcre2 checks to use pkg-config, which should fix #4711 
* Prefers pcre2 over pcre if both are available

Fixes: #3861 #4711

### CODE REVIEW

**Comment by elliefm** [comment]
@elliefm When backporting, leave out the "prefer pcre2..." commit

**Review by rsto — APPROVED** [review]
Builds and unit tests clean on my build

    "pcre": false,
    "pcre2": true,

**Comment by ksmurchison** [comment]
> Builds and unit tests clean on my build
> 
> ```
> "pcre": false,
> "pcre2": true,
> ```

Same here.

**Review by ksmurchison — APPROVED** [review]

**Comment by elliefm** [comment]
@rsto I've added upgrade instructions to the changes file to clarify the pcre/pcre2/pcre3 situation, to explain how to make configure use the older one when both are present (otherwise it will prefer the newer one), and to explain how to deal with a situation where pcre2 is half-installed already (but not enough for Cyrus to use).  I think this should address your issues deploying this PR to Fastmail -- in a generic way that will be useful to non-Fastmail deployments.

Sorry I missed this earlier.  My bullseye VM already had `libpcre2-dev` installed as a dependency for some other thing, so in my testing I didn't need to do anything in particular to switch between versions.  It hadn't occurred to me that a system might have enough of pcre2 to detect but not enough of it to actually use.

In Fastmail-specific terms (but please read the generic instructions too): I think you need to modify `mkdebian.pl` to either:
* replace `libpcre3-dev` with `libpcre2-dev` in `Build-Depends:` (as well as whatever else it takes to get that package and its dependencies actually installed onto systems); **OR**
* add `--disable-pcre2` to the `./configure` line under `build:`.

We can get away with staying on the old (unsupported, unmaintained) pcre (aka pcre3) for the time being, but we will be forced to upgrade to pcre2 when we upgrade to Bookworm, so IMO we might as well get it over with now.

**Review by rsto — APPROVED** [review]
Thanks, looks good to me.

**Review by ksmurchison — APPROVED** [review]
LGTM, with a small nit

File: imap/cyr_buildinfo.c [hunk 1]
```diff
@@ -202,6 +202,11 @@ static json_t *buildinfo()
 #else
     json_object_set_new(dependency, "pcre", json_false());
 #endif
+#if defined(ENABLE_REGEX) && defined(HAVE_PCRE2POSIX_H)
```
[the remark this answers is not in the document]
Comment (elliefm) [reply imap/cyr_buildinfo.c:205]: Hmm.  Maybe, though nested ifdefs get cluttered fast.  There's also other possibilities here that cyr_buildinfo doesn't reflect at all (librx, libregex).  I do think it's worth cleaning up properly at some point, but my goal here was just to add pcre2 support alongside what was already there.

**Comment by elliefm** [comment]
@elliefm when you get to backporting this, the 3.4 conversion is already done in #4738 (just needs release notes as usual)

**Comment by elliefm** [comment]
@elliefm and also #4749

**Pull request merged**